File: solutions/haskell/complex-numbers/1/src/ComplexNumbers.hs
module ComplexNumbers
(Complex,
 conjugate,
 abs,
 exp,
 real,
 imaginary,
 mul,
 add,
 sub,
 div,
 complex) where

import Prelude hiding (div, abs, exp)
import qualified Prelude as P

-- Data definition -------------------------------------------------------------
data Complex a = Complex { real :: a, imaginary :: a } deriving(Eq, Show)

complex :: (a, a) -> Complex a
complex (r, i) = Complex { real = r, imaginary = i }

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate n = Complex {real = real n, imaginary = -(imaginary n)}

abs :: Floating a => Complex a -> a
abs (Complex re im) = sqrt ( (re*re) + (im*im) )

exp :: Floating a => Complex a -> Complex a
exp (Complex a b) = Complex {real = ea * cos b, imaginary = ea * sin b}
  where ea = P.exp a

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (Complex a b) (Complex c d) =
  Complex { real = a*c - b*d, imaginary = a*d + b*c }

add :: Num a => Complex a -> Complex a -> Complex a
add (Complex a b) (Complex c d) = Complex{ real = a+c, imaginary = b+d }

sub :: Num a => Complex a -> Complex a -> Complex a
sub (Complex a b) (Complex c d) = Complex{ real = a-c, imaginary = b-d }

div :: Fractional a => Complex a -> Complex a -> Complex a
div (Complex u v) (Complex x y) = Complex{
    real = (u*x + v*y) / absPow2,
    imaginary = (v*x - u*y) / absPow2
    }
    where
      absPow2 = x*x + y*y

File: solutions/erlang/complex-numbers/1/src/complex_numbers.erl
-module(complex_numbers).
-export([abs/1, add/2, conjugate/1, divide/2, equal/2, exp/1, imaginary/1, mul/2, new/2,
	 real/1, sub/2]).
-define(Precision, 0.00000001).
-record(complex, {re, im}).

new(Re, Im) -> #complex{re = Re, im = Im}.

real(#complex{re=X}) -> X.

imaginary(#complex{im=X}) -> X.

equal(#complex{re=Re1, im=Im1}, #complex{re=Re2, im=Im2}) when Re1 == Re2, Im1 == Im2 -> true;
equal(#complex{re=Re1, im=Im1}, #complex{re=Re2, im=Im2}) ->
  (erlang:abs(Re1 - Re2) < ?Precision) and (erlang:abs(Im1 - Im2) < ?Precision).

% unary operators

abs(#complex{re=Re, im=Im}) -> math:sqrt(Re*Re + Im*Im).

conjugate(Z) -> Z#complex{im = -Z#complex.im}.

% binary operators

add(#complex{re=R1, im=I1}, #complex{re=R2, im=I2}) -> new(R1+R2, I1+I2).

% (a + bi)(c + di) = (ac - bd) + (bc + ad)i
mul(#complex{re=R1, im=I1}, #complex{re=R2, im=I2}) -> new(R1*R2 - I1*I2, R1*I2 + R2*I1).

sub(#complex{re=R1, im=I1}, #complex{re=R2, im=I2}) -> new(R1-R2, I1-I2).

% (a + bi) / (c + di) = (ac + bd) / (c^2 + d^2) + (bc - ad) / (c^2 + d^2)i
divide(#complex{re=Re1, im=Im1}, #complex{re=Re2, im=Im2}) ->
  B = Re2*Re2 + Im2*Im2,
  new( (Re1*Re2 + Im1*Im2)/B, (Re2*Im1 - Re1*Im2)/B) .

% e^(a + bi) = e^a*e^(bi) = e^a(cos(b) + isin(b))
exp(#complex{re=A, im=B}) ->
  EA = math:exp(A),
  new(EA*math:cos(B), EA*math:sin(B)).

